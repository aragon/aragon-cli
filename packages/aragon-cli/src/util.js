const findUp = require('find-up')
const path = require('path')
const execa = require('execa')
const net = require('net')
const pathToPackage = require('global-modules-path').getPath
const { getInstalledPathSync } = require('get-installed-path')

let cachedProjectRoot

const PGK_MANAGER_BIN_NPM = 'npm'

const findProjectRoot = () => {
  if (!cachedProjectRoot) {
    try {
      cachedProjectRoot = path.dirname(findUp.sync('arapp.json'))
    } catch (_) {
      throw new Error('This directory is not an Aragon project')
      // process.exit(1)
    }
  }
  return cachedProjectRoot
}

const isPortTaken = async (port, opts) => {
  opts = Object.assign({ timeout: 1000 }, opts)

  return new Promise(resolve => {
    const socket = new net.Socket()

    const onError = () => {
      socket.destroy()
      resolve(false)
    }

    socket.setTimeout(opts.timeout)
    socket.on('error', onError)
    socket.on('timeout', onError)

    socket.connect(
      port,
      opts.host,
      () => {
        socket.end()
        resolve(true)
      }
    )
  })
}

const getNodePackageManager = () => {
  return PGK_MANAGER_BIN_NPM
}

const installDeps = (cwd, task) => {
  const bin = getNodePackageManager()
  const installTask = execa(bin, ['install'], { cwd })
  installTask.stdout.on('data', log => {
    if (!log) return
    task.output = log
  })

  return installTask.catch(err => {
    throw new Error(
      `${err.message}\n${
        err.stderr
      }\n\nFailed to install dependencies. See above output.`
    )
  })
}

const getNPMBinary = (packageName, relativeBinaryPath) => {
  let binaryPath
  try {
    binaryPath = `${path.join(
      getInstalledPathSync(packageName, { local: true }),
      relativeBinaryPath
    )}`
  } catch (e) {
    binaryPath = `${path.join(
      pathToPackage('@aragon/cli'),
      'node_modules',
      packageName,
      path.normalize(relativeBinaryPath)
    )}`
  }
  return binaryPath
}

const getContract = (pkg, contract) => {
  const artifact = require(`${pkg}/build/contracts/${contract}.json`)
  return artifact
}

const ANY_ENTITY = '0xffffffffffffffffffffffffffffffffffffffff'

module.exports = {
  findProjectRoot,
  isPortTaken,
  installDeps,
  getNodePackageManager,
  getNPMBinary,
  getContract,
  ANY_ENTITY,
}
