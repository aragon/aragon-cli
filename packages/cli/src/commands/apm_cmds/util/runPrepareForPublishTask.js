import tmp from 'tmp-promise'
import TaskList from 'listr'
import path from 'path'
import { readJsonSync, readFile } from 'fs-extra'
import {
  APM_INITIAL_VERSIONS,
  generateApplicationArtifact,
  loadArappFile,
  getApm,
  useEnvironment,
} from '@aragon/toolkit'
//
import listrOpts from '../../../helpers/listr-options'
import askToConfirm from '../../../lib/publish/askToConfirm'
import { prepareFilesForPublishing } from '../../../lib/publish/preprareFiles'
import {
  changeManifestForHttpServedFrom,
  writeApplicationArtifact,
  flattenCodeFileExists,
  generateFlattenedCode,
  artifactExists,
  checkIfNewArticatIsIdentical,
  copyCurrentApplicationArtifacts,
} from '../../../lib/apm/generateArtifact'

/**
 * ctx mandatory output
 * - pathToPublish {string}
 * @return {TaskList} Tasks
 */
export default async function runPrepareForPublishTask({
  // Globals
  cwd,
  environment,

  // Arguments

  /// Files
  publishDir,
  files,
  ignore,

  /// Http
  httpServedFrom,

  /// Storage
  provider,

  /// Conditionals
  onlyArtifacts,
  onlyContent,
  http,

  // Context
  initialRepo,
  initialVersion,
  version,
  contractAddress,
  silent,
  debug,
}) {
  const { web3, apmOptions, appName } = useEnvironment(environment)

  const apm = await getApm(environment)

  const arapp = loadArappFile()

  const { path: contractPath, roles } = arapp

  return new TaskList(
    [
      {
        title: 'Prepare files for publishing',
        enabled: () => !http,
        task: async ctx => {
          // If no publishDir, create temporary directory
          const pathToPublish = publishDir || (await tmp.dir()).path

          await prepareFilesForPublishing(pathToPublish, files, ignore)

          // (TODO): ctx mutation
          ctx.pathToPublish = pathToPublish

          return `Files copied to temporary directory: ${pathToPublish}`
        },
      },
      {
        title:
          'Check for --http-served-from argument and copy manifest.json to destination',
        enabled: () => http,
        task: async ctx => {
          if (!httpServedFrom)
            throw new Error('You need to provide --http-served-from argument')

          changeManifestForHttpServedFrom(httpServedFrom)

          // (TODO): ctx mutation
          ctx.pathToPublish = httpServedFrom
        },
      },
      {
        title: 'Generate application artifact',
        skip: () => onlyContent && !contractPath,
        task: async (ctx, task) => {
          const outputPath = onlyArtifacts ? cwd : ctx.pathToPublish

          /**
           * Define `performArtifcatGeneration` and `invokeArtifactGeneration`
           * previously so they can be reused below depending on what question
           * has to be asked to the user.
           * NOTE: `listr-input` (wrapped in askToConfirm) requires to be
           * returned in order to work, so it's necessary to nest callbacks
           *
           * TODO: (Gabi) Use inquier to handle confirmation
           */

          async function performArtifcatGeneration(artifact) {
            writeApplicationArtifact(artifact, outputPath)
            await generateFlattenedCode(outputPath, contractPath)
            return `Saved artifact in ${outputPath}`
          }

          async function invokeArtifactGeneration() {
            const contractInterfacePath = path.resolve(
              cwd,
              'build/contracts',
              path.basename(contractPath, '.sol') + '.json'
            )

            const contractInterface = readJsonSync(contractInterfacePath)
            const sourceCode = await readFile(contractPath, 'utf8')

            const artifact = await generateApplicationArtifact(
              arapp,
              contractInterface.abi,
              sourceCode
            )
            performArtifcatGeneration(artifact)
          }

          // If an artifact file exist we check it to reuse
          if (artifactExists(outputPath)) {
            const currentArtifactIsIdentical = !checkIfNewArticatIsIdentical(
              cwd,
              roles,
              contractPath,
              outputPath
            )
            if (!currentArtifactIsIdentical) {
              return askToConfirm(
                "Couldn't reuse artifact due to mismatches, regenerate now?",
                invokeArtifactGeneration
              )
            } else {
              return task.skip('Using existing artifact')
            }
          }

          // If only content we fetch artifacts from previous version
          if (onlyContent & !APM_INITIAL_VERSIONS.includes(version)) {
            try {
              task.output = 'Fetching artifacts from previous version'
              await copyCurrentApplicationArtifacts(
                web3,
                cwd,
                outputPath,
                initialRepo,
                version,
                roles,
                contractPath,
                apmOptions
              )
              if (!flattenCodeFileExists(outputPath)) {
                await generateFlattenedCode(outputPath, contractPath)
              }
              return task.skip(`Using artifacts from v${initialVersion}`)
            } catch (e) {
              if (e.message === 'Artifact mismatch') {
                return askToConfirm(
                  "Couldn't reuse existing artifact due to mismatches, regenerate now?",
                  invokeArtifactGeneration
                )
              } else {
                return askToConfirm(
                  "Couldn't fetch current artifact version to copy it. Please make sure your IPFS or HTTP server are running. Otherwise, generate now?",
                  invokeArtifactGeneration
                )
              }
            }
          }

          return invokeArtifactGeneration()
        },
      },
      {
        title: `Publish intent`,
        task: async (ctx, task) => {
          ctx.contractInstance = null // clean up deploy sub-command artifacts
          const accounts = await web3.eth.getAccounts()
          const from = accounts[0]
          ctx.intent = await apm.publishVersionIntent(
            from,
            appName,
            version,
            http ? 'http' : provider,
            http || ctx.pathToPublish,
            contractAddress
          )
        },
      },
    ],
    listrOpts(silent, debug)
  ).run()
}
