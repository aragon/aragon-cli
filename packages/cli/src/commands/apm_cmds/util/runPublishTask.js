import TaskList from 'listr'
import { getApm, useEnvironment } from '@aragon/toolkit'
//
import { task as execTask } from '../../../commands/dao_cmds/utils/execHandler'
import listrOpts from '../../../helpers/listr-options'

export default async function runPublishTask({
  // Globals
  environment,
  http,
  provider,

  // Arguments
  /// Conditionals
  onlyArtifacts,

  /// Context
  version,
  pathToPublish,
  contractAddress,
  dao,
  proxyAddress,
  methodName,
  params,
  silent,
  debug,
}) {
  const { web3, gasPrice, appName } = useEnvironment(environment)
  // TODO: Move web3 logic inside toolkit

  const apm = await getApm(environment)

  return new TaskList(
    [
      {
        title: `Publish ${appName}`,
        enabled: () => !onlyArtifacts,
        task: async (ctx, task) => {
          if ((environment === 'local') | (environment === 'default')) {
            ctx.contractInstance = null // clean up deploy sub-command artifacts

            task.output = 'Generating transaction and waiting for confirmation'
            const accounts = await web3.eth.getAccounts()
            const from = accounts[0]

            const transaction = await apm.publishVersion(
              from,
              appName,
              version,
              http ? 'http' : provider,
              http || pathToPublish,
              contractAddress,
              from
            )

            transaction.from = from
            transaction.gasPrice = gasPrice
            // apm.js already calculates the recommended gas

            ctx.receipt = await web3.eth.sendTransaction(transaction)
          } else {
            return execTask({
              environment,
              dao,
              app: proxyAddress,
              method: methodName,
              params,
            })
          }
        },
      },
    ],
    listrOpts(silent, debug)
  ).run()
}
