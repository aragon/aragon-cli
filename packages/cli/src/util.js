import execa from 'execa'
import fs from 'fs'
import os from 'os'
import { readJson } from 'fs-extra'
import { request } from 'http'
import inquirer from 'inquirer'
import { getNodePackageManager } from './lib/node'

/**
 * Some characters are rendered differently depending on the OS.
 *
 * @param {string} stdout
 */
export function normalizeOutput(stdout) {
  const next = stdout
    // remove user-specific paths
    .replace(/❯/g, '>')
    .replace(/ℹ/g, 'i')
    // TODO: remove after https://github.com/aragon/aragon-cli/issues/367 is fixed
    .replace(/cli.js/g, 'aragon')
    .replace(/index.js/g, 'aragon')
    // replace homedir in paths
    .replace(new RegExp(os.homedir(), 'g'), '~')
    // sometimes there's an extra LF
    .trim()

  return next
}

/**
 * Check if an http server is listening at a given url
 * @param {string} url Server url
 * @returns {boolean} true if server is returning a valid response
 */
export function isHttpServerOpen(url) {
  return new Promise(resolve => {
    request(url, { method: 'HEAD' }, r => {
      resolve(r.statusCode >= 200 && r.statusCode < 400)
    })
      .on('error', () => resolve(false))
      .end()
  })
}

export const installDeps = (cwd, task) => {
  const bin = getNodePackageManager()
  const installTask = execa(bin, ['install'], { cwd })
  installTask.stdout.on('data', log => {
    if (!log) return
    task.output = log
  })

  return installTask.catch(err => {
    throw new Error(
      `${err.message}\n${err.stderr}\n\nFailed to install dependencies. See above output.`
    )
  })
}

/**
 * Run a npm script.
 * @param {string} scriptName
 * @return {Promise<string>} feeback, if the helper aborts early it may return
 * a string with a reason message
 */
export const runScriptHelper = async (scriptName, logCallback) => {
  if (!fs.existsSync('package.json')) {
    return 'No package.json found'
  }

  const { scripts = {} } = await readJson('package.json')
  if (!scripts[scriptName]) {
    return `${scriptName} script not defined in package.json`
  }

  const bin = getNodePackageManager()
  const scriptProcess = execa(bin, ['run', scriptName])

  scriptProcess.stdout.on('data', log => {
    if (logCallback) logCallback(`npm run ${scriptName}: ${log}`)
  })

  try {
    await scriptProcess
  } catch (err) {
    throw new Error(
      `${err.message}\n${err.stderr}\n\nScript ${scriptName} failed. See above output.`
    )
  }
}

/**
 * Parse a String to Boolean, or throw an error.
 *
 * The check is **case insensitive**! (Passing `"TRue"` will return `true`)
 *
 * @param {string} target must be a string
 * @returns {boolean} the parsed value
 */
export const parseAsBoolean = target => {
  if (typeof target !== 'string') {
    throw new Error(
      `Expected ${target} to be of type string, not ${typeof target}`
    )
  }

  const lowercase = target.toLowerCase()

  if (lowercase === 'true') {
    return true
  }

  if (lowercase === 'false') {
    return false
  }

  throw new Error(`Cannot parse ${target} as boolean`)
}

/**
 * Parse a String to Array, or throw an error.
 *
 * @param {string} target must be a string
 * @returns {Array} the parsed value
 */
export const parseAsArray = target => {
  if (typeof target !== 'string') {
    throw new Error(
      `Expected ${target} to be of type string, not ${typeof target}`
    )
  }

  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse
  const json = JSON.parse(target)

  if (Array.isArray(json)) {
    return json
  }

  throw new Error(`Cannot parse ${target} as array`)
}

/**
 * Parse a String to Boolean or Array, or throw an error.
 *
 * @param {string} target must be a string
 * @returns {boolean|Array} the parsed value
 */
export const parseArgumentStringIfPossible = target => {
  // convert to boolean: 'false' to false
  try {
    return parseAsBoolean(target)
  } catch (e) {}

  // convert to array: '["hello", 1, "true"]' to ["hello", 1, "true"]
  // TODO convert children as well ??
  try {
    return parseAsArray(target)
  } catch (e) {}

  // nothing to parse
  return target
}

export const askForInput = async message => {
  const { reply } = await inquirer.prompt([
    {
      type: 'input',
      name: 'reply',
      message,
    },
  ])
  return reply
}

export const askForChoice = async (message, choices) => {
  const { reply } = await inquirer.prompt([
    {
      type: 'list',
      name: 'reply',
      message,
      choices,
    },
  ])
  return reply
}

export const askForConfirmation = async message => {
  const { reply } = await inquirer.prompt([
    {
      type: 'confirm',
      name: 'reply',
      message,
    },
  ])
  return reply
}
