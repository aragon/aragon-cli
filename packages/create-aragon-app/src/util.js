const execa = require('execa')

const PGK_MANAGER_BIN_NPM = 'npm'

const getNodePackageManager = () => {
  return PGK_MANAGER_BIN_NPM
}

const installDeps = (cwd, task) => {
  const bin = getNodePackageManager()
  const installTask = execa(bin, ['install'], { cwd })

  installTask.stdout.on('data', bytes => {
    const str = Buffer.from(bytes, 'utf-8').toString()
    task.output = str
  })

  return installTask.catch(err => {
    throw new Error(
      `${err.message}\n${err.stderr}\n\nFailed to install dependencies. See above output.`
    )
  })
}

/**
 * Validates an Aragon Id
 * @param {string} aragonId Aragon Id
 * @returns {boolean} `true` if valid
 */
function isValidAragonId(aragonId) {
  return /^[a-z0-9-]+$/.test(aragonId)
}

module.exports = {
  installDeps,
  isValidAragonId,
  getNodePackageManager,
}
