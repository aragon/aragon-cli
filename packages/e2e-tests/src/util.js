import execa from 'execa'
import psTree from 'ps-tree'

// let's try gracefully, otherwise we can do SIGTERM or SIGKILL
const defaultKillSignal = 'SIGINT'
const defaultLogger = process.env.DEBUG ? console.log : () => { }

export async function startBackgroundProcess ({
  cmd,
  args,
  execaOpts,
  readyOutput,
  logger = defaultLogger,
  killSignal = defaultKillSignal
}) {
  return new Promise((resolve, reject) => {
    // start the process
    const subprocess = execa(cmd, args, execaOpts)
    logger(cmd, 'spawned with PID: ', subprocess.pid)
    let stdout = ''
    let stderr = ''

    // return this function so the process can be killed
    const exit = () => new Promise((resolve, reject) => {
      psTree(subprocess.pid, (err, children) => {
        if (err) reject(err)

        children.map(child => { // each child has the properties: COMMAND, PPID, PID, STAT
          logger(cmd, 'killing child: ', child)
          process.kill(child.PID, killSignal)
        })

        resolve()
      })
    })

    subprocess.stdout.on('data', data => {
      // parse
      data = data.toString()
      // log
      logger(cmd, 'stdout:', data)
      // build output stream
      stdout += data
      // check for ready signal
      if (data.includes(readyOutput)) {
        resolve({
          exit,
          stdout
        })
      }
    })

    subprocess.stderr.on('data', data => {
      // parse
      data = data.toString()
      // log
      logger(cmd, 'stderr:', data)
      // build error stream
      stderr += data
    })

    subprocess.on('close', (code, signal) => {
      // log
      logger(cmd, 'closing with code:', code, 'and signal:', signal)

      // reject only if the promise did not previously resolve 
      // which means this is probably getting killed by the test which is ok
      if (!stdout.includes(readyOutput)) {
        const err = new Error(`Process closed unexpectedly with code ${code} and signal ${signal}`)
        err.stdout = stdout
        err.stderr = stderr
        reject(err)
      }
    })

    subprocess.on('exit', (code, signal) => {
      logger(cmd, 'exiting with code:', code, 'and signal:', signal)
    })
  })
}

/**
 * Some characters are rendered differently depending on the OS.
 * 
 * @param {string} stdout 
 */
export function normalizeOutput (stdout) {
  const next = stdout
    .replace(/❯/g, '>')
    .replace(/ℹ/g, 'i')
    // TODO: remove after https://github.com/aragon/aragon-cli/issues/367 is fixed 
    .replace(/cli.js/g, 'aragon')
    // sometimes there's an extra LF
    .trim()

  return next
}
