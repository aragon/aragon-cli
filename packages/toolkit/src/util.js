import path from 'path'
import findUp from 'find-up'
//
import {
  ARAGON_DOMAIN,
  DEFAULT_GAS_FUZZ_FACTOR,
  LAST_BLOCK_GAS_LIMIT_FACTOR,
} from './helpers/constants'

let cachedProjectRoot

export const findProjectRoot = () => {
  if (!cachedProjectRoot) {
    try {
      cachedProjectRoot = path.dirname(findUp.sync('arapp.json'))
    } catch (_) {
      throw new Error('This directory is not an Aragon project')
    }
  }
  return cachedProjectRoot
}

/**
 * Check eth address equality without checksums
 * @param {string} first address
 * @param {string} second address
 * @returns {boolean} address equality
 */
export function addressesEqual(first, second) {
  first = first && first.toLowerCase()
  second = second && second.toLowerCase()
  return first === second
}

export const isAddress = addr => /0x[a-fA-F0-9]{40}/.test(addr)

/**
 * Validates an Aragon Id
 * @param {string} aragonId Aragon Id
 * @returns {boolean} `true` if valid
 */
export function isValidAragonId(aragonId) {
  return /^[a-z0-9-]+$/.test(aragonId)
}

/**
 * Convert a DAO id to its subdomain
 * E.g. mydao -> mydao.aragonid.eth
 * @param {string} aragonId Aragon Id
 * @returns {string} DAO subdomain
 */
export function convertDAOIdToSubdomain(aragonId) {
  // If already a subdomain, return
  if (new RegExp(`^([a-z0-9-]+).${ARAGON_DOMAIN}$`).test(aragonId))
    return aragonId

  if (!isValidAragonId(aragonId)) throw new Error(`Invalid DAO Id: ${aragonId}`)

  return `${aragonId}.${ARAGON_DOMAIN}`
}

export const expandLink = link => {
  const { name, address } = link
  const placeholder = `__${name}${'_'.repeat(38 - name.length)}`
  link.placeholder = placeholder
  link.regex = new RegExp(placeholder, 'g')
  link.addressBytes = address.slice(0, 2) === '0x' ? address.slice(2) : address
  return link
}

/**
 *
 * Calculate the recommended gas limit
 *
 * @param {*} web3 eth provider to get the last block gas limit
 * @param {number} estimatedGas estimated gas
 * @param {number} gasFuzzFactor defaults to 1.5
 * @returns {number} gasLimit
 */
export const getRecommendedGasLimit = async (
  web3,
  estimatedGas,
  gasFuzzFactor = DEFAULT_GAS_FUZZ_FACTOR
) => {
  // TODO print these values if --debug is passed
  const latestBlock = await web3.eth.getBlock('latest')
  const blockGasLimit = latestBlock.gasLimit

  const upperGasLimit = Math.round(blockGasLimit * LAST_BLOCK_GAS_LIMIT_FACTOR)
  if (estimatedGas > upperGasLimit) return estimatedGas // TODO print a warning?

  const bufferedGasLimit = Math.round(estimatedGas * gasFuzzFactor)

  if (bufferedGasLimit < upperGasLimit) return bufferedGasLimit
  return upperGasLimit
}

export const getContract = (pkg, contract) => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const artifact = require(`${pkg}/build/contracts/${contract}.json`)
  return artifact
}

export const getIpfsCacheFiles = () => {
  return path.resolve(require.resolve('@aragon/aragen'), '../ipfs-cache')
}

export const getAragonGanacheFiles = () => {
  return path.resolve(require.resolve('@aragon/aragen'), '../aragon-ganache')
}
