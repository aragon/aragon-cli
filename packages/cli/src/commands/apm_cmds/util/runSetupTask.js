import TaskList from 'listr'
import { isAddress } from 'web3-utils'
import {
  ZERO_ADDRESS,
  APM_INITIAL_VERSIONS,
  isLocalDaemonRunning,
  startLocalDaemon,
  getBinaryPath,
  getDefaultRepoPath,
  getApmRepo,
  useEnvironment,
} from '@aragon/toolkit'

// helpers
import { compileContracts } from '../../../helpers/truffle-runner'
import listrOpts from '../../../helpers/listr-options'

// cmds
import {
  getPrevAndNextVersion,
  InvalidBump,
} from '../../../lib/apm/getPrevAndNextVersion'
import { task as deployTask } from '../../deploy'

// util
import { runScriptHelper } from '../../../util'

/**
 * @typedef {Object} VersionAppInfo
 * @property {string} contractAddress "0xaabb1122"
 * @property {string} version "0.2.4"
 * // FUN! may include:
 * { error, contentURI }
 * or
 * { manifest, module, { content: { provider, location } }
 * or
 * { content: { provider, location } }
 *
 * // Where: if contentURI = "ipfs:Qmazz"
 * - provider: "ipfs"
 * - location: "Qmazz"
 */

/**
 * ctx mandatory output:
 * - initialRepo {VersionAppInfo}
 * - initialVersion {string} Version before release: "0.2.4"
 * - version {string} To release version: "0.2.5"
 * @return {TaskList} Tasks
 */
export default async function runSetupTask({
  // Globals
  cwd,
  environment,

  // Arguments

  /// Scritps
  prepublish,
  prepublishScript,
  build,
  buildScript,

  /// Version
  bump: bumpOrVersion,

  /// Contract
  contract,
  init,
  reuse,

  /// Conditionals
  onlyContent,
  onlyArtifacts,
  http,
  silent,
  debug,
}) {
  const { appName } = useEnvironment(environment)

  if (onlyContent) {
    contract = ZERO_ADDRESS
  }

  // Set prepublish to true if --prepublish-script argument is used
  if (process.argv.includes('--prepublish-script')) prepublish = true

  /**
   * Flag for the Deploy contract task
   * @type {boolean}
   */
  let shouldDeployContract

  return new TaskList(
    [
      {
        title: 'Start IPFS',
        skip: async () => isLocalDaemonRunning(),
        task: async () => {
          await startLocalDaemon(getBinaryPath(), getDefaultRepoPath(), {
            detached: false,
          })
        },
      },
      {
        // TODO: During test decide which prepublish script choose here and for building
        title: 'Running prepublish script',
        enabled: () => prepublish,
        task: async (_0, task) => {
          const abortReason = await runScriptHelper(prepublishScript, log => {
            task.output = log
          })
          if (abortReason) task.skip(abortReason)
        },
      },
      {
        title: `Applying version bump (${bumpOrVersion})`,
        task: async (ctx, task) => {
          task.output = 'Fetching latest version from aragonPM...'
          try {
            const {
              initialRepo,
              prevVersion,
              version,
              shouldDeployContract: _shouldDeployContract,
            } = await getPrevAndNextVersion(bumpOrVersion, environment)

            // (TODO): For now MUST be exposed in the context because their are used around
            ctx.initialRepo = initialRepo
            ctx.initialVersion = prevVersion
            ctx.version = version
            shouldDeployContract = _shouldDeployContract
          } catch (e) {
            if (e instanceof InvalidBump)
              throw Error(
                "Version bump is not valid, you have to respect APM's versioning policy.\nCheck the version upgrade rules in the documentation:\n  https://hack.aragon.org/docs/apm-ref.html#version-upgrade-rules"
              )
          }
        },
      },
      {
        title: 'Building frontend',
        enabled: () => build && !http,
        task: async (_0, task) => {
          const abortReason = await runScriptHelper(buildScript, log => {
            task.output = log
          })
          if (abortReason) task.skip(abortReason)
        },
      },
      {
        title: 'Compile contracts',
        enabled: () => !onlyContent && isAddress(contract),
        task: async () => compileContracts(),
      },
      {
        title: 'Deploy contract',
        enabled: () =>
          !onlyContent &&
          ((contract && !isAddress(contract)) ||
            (!contract && shouldDeployContract && !reuse)),
        task: async () => {
          return deployTask({
            cwd,
            environment,
            contract,
            init,
          })
        },
      },
      {
        title: 'Determine contract address for version',
        enabled: () => !onlyArtifacts,
        task: async (ctx, task) => {
          if (isAddress(contract)) {
            ctx.contract = contract
            return `Using ${contract}`
          }

          // Get address of deployed contract
          ctx.contract = ctx.contractAddress

          if (!ctx.contract && !APM_INITIAL_VERSIONS.includes(ctx.version)) {
            // Check if we can fall back to a previous contract address
            task.output = 'No contract address provided, using previous one'

            try {
              const { contractAddress } = await getApmRepo(
                appName,
                'latest',
                environment
              )
              ctx.contract = contractAddress
              return `Using ${contractAddress}`
            } catch (err) {
              throw new Error('Could not determine previous contract')
            }
          }

          // Contract address required for initial version
          if (!ctx.contract) {
            throw new Error('No contract address supplied for initial version')
          }

          return `Using ${ctx.contractAddress}`
        },
      },
    ],
    listrOpts(silent, debug)
  ).run()
}
