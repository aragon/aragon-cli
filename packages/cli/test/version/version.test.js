import { runAragonCLI } from '../util'

jest.setTimeout(60000)

test('should return the correct version', async () => {
  const result = await runAragonCLI(['--version'])

  delete result.stdout

  expect(result).toMatchSnapshot()
})
