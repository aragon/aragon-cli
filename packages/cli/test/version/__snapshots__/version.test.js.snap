// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the correct version 1`] = `
Object {
  "all": undefined,
  "command": "node dist/index.js --version",
  "exitCode": 0,
  "failed": false,
  "isCanceled": false,
  "killed": false,
  "stderr": "",
  "timedOut": false,
}
`;
